### PR TITLE
config refactor

### PR DESCRIPTION
**Pull request by qibobo**
1 Refactor config and testcase
2 Aggregator will set triggers to AppEvaluationManager

### CODE REVIEW

File: src/eventgenerator/aggregator/aggregator.go [hunk 1]
```diff
@@ -5,6 +5,7 @@ import (
 	"code.cloudfoundry.org/clock"
 	"code.cloudfoundry.org/lager"
 	"db"
+	. "eventgenerator/generator"
```
Comment (fraenkel) [inline src/eventgenerator/aggregator/aggregator.go]: Do not dot import

File: src/eventgenerator/aggregator/aggregator.go [hunk 2]
```diff
@@ -48,15 +51,31 @@ func (a *Aggregator) ConsumePolicy(policyMap map[string]*model.Policy, appChan c
...
 	}
+	var triggerArrayMap map[string][]*model.Trigger = make(map[string][]*model.Trigger)
 	for appId, policy := range policyMap {
 		for _, rule := range policy.TriggerRecord.ScalingRules {
 			appChan <- &model.AppMonitor{
 				AppId:      appId,
 				MetricType: rule.MetricType,
 				StatWindow: rule.StatWindow,
 			}
+			_, exist := triggerArrayMap[appId+"#"+rule.MetricType]
```
Comment (fraenkel) [inline src/eventgenerator/aggregator/aggregator.go]: Create the string once.

File: src/eventgenerator/aggregator/aggregator.go [hunk 3]
```diff
@@ -48,15 +51,31 @@ func (a *Aggregator) ConsumePolicy(policyMap map[string]*model.Policy, appChan c
...
+			if !exist {
+				triggerArrayMap[appId+"#"+rule.MetricType] = []*model.Trigger{}
+			}
+			triggerArrayMap[appId+"#"+rule.MetricType] = append(triggerArrayMap[appId+"#"+rule.MetricType], &model.Trigger{
```
Comment (fraenkel) [inline src/eventgenerator/aggregator/aggregator.go]: This is wasteful.
You got it once and decided not to hold on to the object and then you get it again.
You should be able to do one retrieve and one put.

File: src/eventgenerator/config/config.go [hunk 4]
```diff
@@ -8,41 +8,144 @@ import (
 )
 
 const (
-	DefaultLoggingLevel               = "info"
-	DefaultPollInterval time.Duration = 30 * time.Second
+	DefaultServerPort                int           = 8080
+	DefaultLoggingLevel                            = "info"
```
Comment (fraenkel) [inline src/eventgenerator/config/config.go]: Why do we have both default constants and default structs?

File: src/eventgenerator/config/config.go [hunk 5]
```diff
@@ -8,41 +8,144 @@ import (
...
+
+var defaultAggregatorConfig = AggregatorConfig{
+	AggregatorExecuateInterval: DefaultAggregatorExecuteInterval,
+	PolicyPollerInterval:       DefaultPolicyPollerInterval,
+	MetricPollerCount:          DefaultMetricPollerCount,
+	AppMonitorChannelSize:      DefaultAppMonitorChannelSize,
+}
+
+type EvaluatorConfig struct {
+	EvaluationManagerIntervalInSeconds int `yaml:"evaluation_manager_execute_interval_in_seconds"`
```
Comment (fraenkel) [inline src/eventgenerator/config/config.go]: These should be durations

File: src/eventgenerator/config/config.go [hunk 6]
```diff
@@ -8,41 +8,144 @@ import (
...
+	Level: DefaultLoggingLevel,
+}
+
+type DBConfig struct {
+	PolicyDBUrl    string `yaml:"policy_db_url"`
+	AppMetricDBUrl string `yaml:"app_metrics_db_url"`
+}
+
+type AggregatorConfig struct {
+	AggregatorExecuateIntervalInSeconds int `yaml:"aggregator_execute_interval_in_seconds"`
```
Comment (fraenkel) [inline src/eventgenerator/config/config.go]: Durations

File: src/eventgenerator/config/config.go [hunk 7]
```diff
@@ -8,41 +8,136 @@ import (
...
+	Level string `yaml:"level"`
+}
+
+type DBConfig struct {
+	PolicyDBUrl    string `yaml:"policy_db_url"`
+	AppMetricDBUrl string `yaml:"app_metrics_db_url"`
+}
+
+type AggregatorConfig struct {
+	AggregatorExecuateIntervalInSeconds time.Duration `yaml:"aggregator_execute_interval_in_seconds"`
```
Comment (fraenkel) [inline src/eventgenerator/config/config.go]: This should now be AggregatorExecuateInterval
They can write 10s or whatever units.

File: src/eventgenerator/config/config.go [hunk 8]
```diff
@@ -8,41 +8,136 @@ import (
...
 	}
 	err := yaml.Unmarshal(bytes, &conf)
 	if err != nil {
 		return nil, err
 	}
-	conf.LogLevel = strings.ToLower(conf.LogLevel)
-
+	conf.Logging.Level = strings.ToLower(conf.Logging.Level)
+	if conf.Aggregator.AggregatorExecuateIntervalInSeconds != 0 {
+		conf.Aggregator.AggregatorExecuateInterval = time.Duration(conf.Aggregator.AggregatorExecuateIntervalInSeconds) * time.Second
```
Comment (fraenkel) [inline src/eventgenerator/config/config.go]: This is wrong given its already a Duration

**Comment by qibobo** [comment]
updated and rebase with the newest develop

**Comment by fraenkel** [comment]
Squashed and merged

**Pull request closed without merging**